src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -160,194 +166,183 @@ impl NonOverlapSubLevelPicker {
+                select_sst_id_set.insert(other.sst_id);
+                all_file_size += other.file_size;
             }
 
             if pending_compact {
                 break;
             }
 
-            // check reverse overlap
-            for reverse_index in (0..last_level_index).rev() {
-                let target_tables = &l0[reverse_index].table_infos;
+            let mut extra_overlap_levels = vec![];
+
+            for reverse_index in (0..target_index).rev() {
+                let target_tables = &levels[reverse_index].table_infos;
+                // It has select all files in this sub-level, so it can not overlap with more files.
+                if ret.sstable_infos[reverse_index].len() == target_tables.len() {
+                    continue;
+                }
                 let overlap_files = overlap_info.check_multiple_overlap(target_tables);
                 let mut extra_overlap_sst = Vec::with_capacity(overlap_files.len());
                 for other in overlap_files {
                     if level_handler.is_pending_compact(&other.sst_id) {
                         pending_compact = true;
                         break;
                     }
-
                     if select_sst_id_set.contains(&other.sst_id) {
                         // Since some of the files have already been selected when selecting
                         // upwards, we filter here to avoid adding sst repeatedly
                         continue;
                     }
 
                     all_file_size += other.file_size;
-                    all_file_count += 1;
-
                     overlap_info.update(&other);
                     select_sst_id_set.insert(other.sst_id);
                     extra_overlap_sst.push(other);
                 }
 
-                if pending_compact {
+                // If the current
+                if extra_overlap_sst.is_empty() {
                     break;
                 }
 
-                if !extra_overlap_sst.is_empty() {
-                    level_select_files[reverse_index].extend(extra_overlap_sst.into_iter());
-                }
+                extra_overlap_levels.push((reverse_index, extra_overlap_sst));
             }
 
+            // check reverse overlap
             if pending_compact {
                 // encountering a pending file means we don't need to continue processing this
                 // interval
                 break;
             }
 
-            if all_file_size < self.min_compaction_bytes || select_level_count < self.min_depth {
-                // Avoiding too small compaction and number of layers
-                continue;
+            if ret
+                .sstable_infos
+                .iter()
+                .filter(|files| !files.is_empty())
+                .count()
+                > self.min_depth
+                && ret.total_file_size > self.min_compaction_bytes
+                && !extra_overlap_levels.is_empty()
```

Comment:
when extra_overlap_levels is not empty, exiting early can cause missed data, the new data is compacted down and the old data (overlap) is read again.